### PR TITLE
Issue #1081: non necessary raw strings

### LINKED ISSUE #1081: Raw strings vs regular strings
One can use both `r'abc'` and `'abc'` to represent a simple string.
But, `r` strings are only required when dealing with `\` in your strings.
So, the rule is: if there's no `\` in your string, then do not use `r`

### PR DESCRIPTION
**Pull request by Mema5**
# First PR for our group
@fwald and I solved this issue today. We added the new rule along with tests and documentation. We also refactored old tests and enforced the new rule in the project. 
Thank you for your answers!

## Checklist
- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues
- Closes #1081 

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

**Comment by sobolevn** [comment]
Wow! @Mema5, @Gwin73 thanks! I will review this shortly.

Offtop: can you please share your story behind this feature? Do you practice pair programming? Or is it a part of some kind of practice? And the most important question for me as a maintainer: why did you choose this project to contribute to? 

You rock! 👾 🚀

**Comment by Mema5** [comment]
Thank you, it is rewarding to receive such a positive feedback :)

Actually we are a group of 5 students (@fward @Gwin73  @fils2 @ruwaid123 @Mema5) and we contribute as part of a course assignment. So yes, it is mostly to learn and gain experience on software engineering practices!
As for the project selection, we spent some time on that and we chose WPS for the following reasons. First, the project was active (thanks to you), the issues seemed interesting and reachable for us. Secondly, the building and testing were relatively easy from the documentation, thanks to poetry. Finally, the members of our team were pleased with the purpose of the project and keen on working in Python :)

The project will only last one week, but we are glad if it is useful!

File: CHANGELOG.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ Semantic versioning in our case means:
 - Forbids to use positional only `/` arguments
 - Adds `__call__` to list of methods that should be on top #1125
 - Now allows `_` to be used as a defined variable
+- Forbids to use raw strings `r''` when not necessary
```
Comment (sobolevn) [inline CHANGELOG.md]: This is related to `0.15`, not `0.14`

File: tests/test_visitors/test_tokenize/test_primitives/test_string_tokens/test_string_modifier.py [hunk 2]
```diff
@@ -43,6 +41,36 @@ def test_correct_prefix(
...
+    'rb',
+    'fr',
+])
+@pytest.mark.parametrize('primitive', [
+    '{0}"\n"',
+    "{0}'\n'",
+    '{0}"Big \ntext"',
+    "{0}'Format \n123'",
+])
+def test_correct_prefix_raw_string(
```
Comment (sobolevn) [inline tests/test_visitors/test_tokenize/test_primitives/test_string_tokens/test_string_modifier.py]: We need `test_wrong_prefix_raw_string` case as well.

File: wemake_python_styleguide/violations/consistency.py [hunk 3]
```diff
@@ -2107,3 +2109,29 @@ class IterableUnpackingViolation(ASTViolation):
...
+
+    Reasoning:
+        Raw string are only needed when dealing with \ in the string.
+
+    Solution:
+        Do not prefix the string with r. Use a normal string instead.
+
+    Example::
+        # Correct:
+        s = r'This is a correct use \'
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py]: `s = ` part can be removed

File: wemake_python_styleguide/violations/consistency.py [hunk 4]
```diff
@@ -2107,3 +2109,29 @@ class IterableUnpackingViolation(ASTViolation):
...
+
+        # Wrong:
+        s = r'This string should not be prefixed with r.'
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py]: The same here

File: wemake_python_styleguide/violations/consistency.py [hunk 5]
```diff
@@ -2107,3 +2109,29 @@ class IterableUnpackingViolation(ASTViolation):
...
+
+    .. versionadded:: 0.14.0
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py]: Version is `0.15.0`

File: wemake_python_styleguide/visitors/tokenize/primitives.py [hunk 6]
```diff
@@ -221,6 +224,15 @@ def _check_wrong_unicode_escape(self, token: tokenize.TokenInfo) -> None:
             # character can never be the start of a new backslash escape.
             index += 2
 
+    def _check_unnecessary_raw_string(self, token: tokenize.TokenInfo) -> None:
+        modifiers, string_def = split_prefixes(token)
+
+        if 'r' in modifiers.lower():
+            if '\\' not in string_def.encode('unicode-escape').decode():
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/tokenize/primitives.py]: `r'\'` is better here.

Comment (sobolevn) [reply wemake_python_styleguide/visitors/tokenize/primitives.py]: Can you please also explain `encode('unicode-escape').decode()` part? Why is it required?

File: wemake_python_styleguide/visitors/tokenize/primitives.py [hunk 7]
```diff
@@ -221,6 +224,15 @@ def _check_wrong_unicode_escape(self, token: tokenize.TokenInfo) -> None:
...
+                self.add_violation(
+                    RawStringNotNeededViolation(token, text=token.string),
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/tokenize/primitives.py]: `text=` can be removed.

File: wemake_python_styleguide/visitors/tokenize/primitives.py (hunk 6 above)
Comment (Mema5) [reply wemake_python_styleguide/visitors/tokenize/primitives.py]: No, `r'\'` is not correct in Python I think.
```console
Python 3.6.9
>>> a=r'\'
  File "<stdin>", line 1
    a=r'\'
         ^
SyntaxError: EOL while scanning string literal
>>> 
```

The `encode('unicode-escape').decode()` is a workaround to convert into a raw string. ([see here](https://stackoverflow.com/a/58309907)). If you know something more beautiful let me know!

Comment (sobolevn) [reply wemake_python_styleguide/visitors/tokenize/primitives.py]: Ok! 👍

File: CHANGELOG.md (hunk 1 above)
Comment (Mema5) [reply CHANGELOG.md]: The header for 0.15 is not yet created. Should I create it or just remove the line of contribution?

Comment (sobolevn) [reply CHANGELOG.md]: You can create it

Comment (Mema5) [reply CHANGELOG.md]: See 870f8bed7a6f92a55ec15e6e2e806d5344041f50

File: tests/test_visitors/test_tokenize/test_primitives/test_string_tokens/test_string_modifier.py (hunk 2 above)
Comment (Mema5) [reply tests/test_visitors/test_tokenize/test_primitives/test_string_tokens/test_string_modifier.py]: See f3b55429539657bced0dd0d376d1efde116ee957

File: wemake_python_styleguide/violations/consistency.py (hunk 3 above)
Comment (Mema5) [reply wemake_python_styleguide/violations/consistency.py]: See c6825cb4e8c8f27480c5fe98bba4ebbea0546b16

File: wemake_python_styleguide/violations/consistency.py (hunk 4 above)
Comment (Mema5) [reply wemake_python_styleguide/violations/consistency.py]: See c6825cb4e8c8f27480c5fe98bba4ebbea0546b16

File: wemake_python_styleguide/violations/consistency.py (hunk 5 above)
Comment (Mema5) [reply wemake_python_styleguide/violations/consistency.py]: See c6825cb4e8c8f27480c5fe98bba4ebbea0546b16

File: wemake_python_styleguide/visitors/tokenize/primitives.py (hunk 7 above)
Comment (Mema5) [reply wemake_python_styleguide/visitors/tokenize/primitives.py]: See c6825cb4e8c8f27480c5fe98bba4ebbea0546b16

File: tests/fixtures/noqa/noqa.py [hunk 8]
```diff
@@ -692,3 +692,5 @@ def consecutive_yields():
     'wrong',
 ]
 *numbers, = [4, 7]  # noqa: WPS356
+
+unnecessary_raw_string = r'this string does not contain any backlashes.' # noqa: WPS357
```
Comment (sobolevn) [inline tests/fixtures/noqa/noqa.py]: New line after `unnecessary_raw_string` is required

File: wemake_python_styleguide/visitors/tokenize/primitives.py (hunk 6 above)
Comment (sobolevn) [reply wemake_python_styleguide/visitors/tokenize/primitives.py]: @Mema5 have you tried running tests without `encode('unicode-escape').decode()`? Did it work?

**Comment by sobolevn** [comment]
@Mema5 can you please rebase your branch?

**Comment by Mema5** [comment]
@sobolevn here it is!

File: wemake_python_styleguide/violations/consistency.py [hunk 9]
```diff
@@ -2143,3 +2145,29 @@ class LineCompriseCarriageReturnViolation(TokenizeViolation):
...
+@final
+class RawStringNotNeededViolation(TokenizeViolation):
+    r"""
+    Forbid the use of raw strings when there is no backslash in the string.
+
+    Reasoning:
+        Raw string are only needed when dealing with \ in the string.
+
+    Solution:
+        Do not prefix the string with r. Use a normal string instead.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py]: Should be:

```
``r``
```

File: wemake_python_styleguide/violations/consistency.py [hunk 10]
```diff
@@ -2143,3 +2145,29 @@ class LineCompriseCarriageReturnViolation(TokenizeViolation):
...
     code = 357
+
+
+@final
+class RawStringNotNeededViolation(TokenizeViolation):
+    r"""
+    Forbid the use of raw strings when there is no backslash in the string.
+
+    Reasoning:
+        Raw string are only needed when dealing with \ in the string.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py]: Should be:

```
``\``
```

File: wemake_python_styleguide/visitors/tokenize/primitives.py [hunk 11]
```diff
@@ -231,6 +233,15 @@ def _check_wrong_unicode_escape(self, token: tokenize.TokenInfo) -> None:
             # character can never be the start of a new backslash escape.
             index += 2
 
+    def _check_unnecessary_raw_string(self, token: tokenize.TokenInfo) -> None:
+        modifiers, string_def = split_prefixes(token)
+
+        if 'r' in modifiers.lower():
+            if '\\' not in string_def:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/tokenize/primitives.py]: Let's combine it into `if 'r' in modifiers.lower() and '\\' not in string_def`

**Pull request merged**